app/validators/efl_year_validator.rb

```diff
@@ -0,0 +1,27 @@
+class EflYearValidator < ActiveModel::EachValidator
+  YEAR_REGEX = /^\d{4}$/.freeze
+
+  def validate_each(record, attribute, value)
+    return if value.blank?
+
+    return record.errors.add(attribute, :not_a_year) if not_a_year(value.to_s)
+    return record.errors.add(attribute, :invalid) if outside_efl_start_years(record, value.to_i)
+    return record.errors.add(attribute, :future) if value.to_i > Time.zone.today.year
+  end
+
+private
+
+  def not_a_year(value)
+    value !~ YEAR_REGEX
+  end
+
+  def outside_efl_start_years(record, value)
+    value < if record.model_name.human == 'Other efl qualification form'
+              1900
+            elsif record.model_name.human == 'Ielts form'
+              1980
+            else
+              1964
+            end
+  end
```

Comment:
Could this be dropped from this validator, which could then only focus on "year", and then specify the range separately?

eg

```ruby
  validates :award_year, presence: true, year: true
  validates :award_year, numericality: { greater_than_or_equal_to: 1990}, if: :ielts?
  validates :award_year, numericality: { greater_than_or_equal_to: 1900}, if: :toefl?
  validates :award_year, numericality: { greater_than_or_equal_to: 1964}, if: :other?
```

Just an idea though!

I also wonder if `numericality: { only_integer: true }` does the same thing in rejecting decimal and exponential numbers? (I only just learnt that was an option: see [numericality](https://guides.rubyonrails.org/active_record_validations.html#numericality))

Reply:
new to me too! top 2 answers on this page both have good bits: https://stackoverflow.com/questions/12611881/best-way-to-validate-year-via-ruby-on-rails-validates-method

Reply:
would be nice to have an interface like `validates year: { after: 1980 }`?

Reply:
Great ideas, thanks! I think I may be able to drop the new validator completely and just use the pre-existing one. It would be something like this on each of the efl forms.

 `validates :award_year, presence: true, 
                                         numericality: { greater_than_or_equal_to: 1980 }, 
                                         year: { future: true }`

This way I can avoid having efl logic in the existing year validator which is generic. Re, the `only_integer: true`. The decimal and exponential fails before validation, i.e no value is sent which is why (i guess) it doesn't retain.

Reply:
@frankieroberto Yep, I've added the `only_integer: true`. As you said, it catches the decimal point/floats and now we can give a more precise error message. Let me know if ok 👍